### ISSUE TITLE
sync not happening when remote repos are referred in ArgoCD using Kustomize

### ISSUE THREAD

**Issue by s-u-b-h-a-k-a-r** [argoproj/argo-cd#4508]
Created base repo:
https://github.com/SubhakarKotta/hello-world-kustomize

Created child repo which uses base:
https://github.com/SubhakarKotta/hello-world-kustomize-preprod

Child repo kustomization.yaml referring the base repo.
 
```yml
apiVersion: kustomize.config.k8s.io/v1beta1
kind: Kustomization
resources:
    - github.com/SubhakarKotta/hello-world-kustomize/base
patchesStrategicMerge:
    - deployment.yaml
```
Argocd:
Create Repository:-
https://github.com/SubhakarKotta/hello-world-kustomize-preprod
Create Application:-

```yml
apiVersion: argoproj.io/v1alpha1
kind: Application
metadata:
  name: hello-kubernetes
spec:
  destination:
    name: ''
    namespace: hello-kubernetes
    server: 'https://kubernetes.default.svc'
  source:
    path: .
    repoURL: 'https://github.com/SubhakarKotta/hello-world-kustomize-preprod.git'
    targetRevision: HEAD
  project: default
  syncPolicy:
    automated:
      prune: true
      selfHeal: true
    syncOptions:
      - CreateNamespace=true
```
Now when any changes happens to 'https://github.com/SubhakarKotta/hello-world-kustomize-preprod.git' autosync happens properly.

Issue:
When the changes are done to base repository the argocd is not able to sync the changes made in base repository.

**Comment by jessesuen** [comment]
This is current design. Argo CD caches the state of manifests by the commit hash (default 24h). If your remote base changes meaning within those 24 hours, it will not be detected unless a force refresh is performed.

We actually discourage this pattern where the remote base can change meaning since it is not gitops friendly (e.g. rollback is no longer deterministic).

**Comment by SubhakarKotta** [comment]
> This is current design. Argo CD caches the state of manifests by the commit hash (default 24h). If your remote base changes meaning within those 24 hours, it will not be detected unless a force refresh is performed.
> 
> We actually discourage this pattern where the remote base can change meaning since it is not gitops friendly (e.g. rollback is no longer deterministic).

Thanks for your reply. Means the base repository sync happens only after 24 hours ?. I tried doing force refresh but still the sync hasn't happened.

**Comment by rumstead** [comment]
I believe you have to force the [repo server to resync](https://argoproj.github.io/argo-cd/user-guide/commands/argocd_repo_get/#options), not a "force" from the UI. My understanding is the [force on the UI](https://argoproj.github.io/argo-cd/user-guide/commands/argocd_app_sync/#options) is a `kubectl apply --force`.

**Comment by ranipm2705** [comment]
@s-u-b-h-a-k-a-r - Did you notice the out-of-sync trigger post 24h for the base repo changes? Or did you try the force-repo-sync?

**Comment by toninis** [comment]
There is another issue open and closed https://github.com/argoproj/argo-cd/issues/1646 . 
You can check [here](https://github.com/argoproj/argo-cd/issues/1646#issuecomment-543949702)

**Comment by lamebear** [comment]
Our use-case is similar but a little bit different. We still have the base and child repo, but our child repo references a specific branch or tag:

```
resources:
    - gitlab.com/path/to/project.git?ref=some-ref
 ```
 
 Our CI constrains that production environments always use a tag, so the rollbacks are always deterministic. The issue comes into play with our development branches. 
 
 What we do for development branches is create a branch in our kustomize base repo that matches the branch in the application repo, say, `feature/some-feature`. Then, in our development kustomize repo, we reference the branch as the ref:
 
 ```
resources:
    - gitlab.com/path/to/project.git?ref=feature/some-feature
 ```

In this case, we don't care about rollbacks, only a forward progression. But, when a new version is deployed to the `feature/some-feature` branch on the kustomize base repo, Argo does not pick it up. My workaround for that is to have the CI create a new branch in the base kustomize repo for every commit and have the development kustomize repo reference the commit branch.

```
resources:
    - gitlab.com/path/to/project.git?ref=feature/some-feature/0072a64
 ```

While this works, this is rather messy IMO and would love the application to update when the base repo updates. That would simplify the CI a bit and not result in so many branches.

**Comment by jdavisp3** [comment]
> This is current design. Argo CD caches the state of manifests by the commit hash (default 24h). If your remote base changes meaning within those 24 hours, it will not be detected unless a force refresh is performed.
> 
> We actually discourage this pattern where the remote base can change meaning since it is not gitops friendly (e.g. rollback is no longer deterministic).

When you say "we discourage this pattern" are you referring to the use of references to other github repos in kustomize files? If so, is there another method you suggest for avoiding a lot of copying between definitions for different environments?

**Comment by rumstead** [comment]
> > This is current design. Argo CD caches the state of manifests by the commit hash (default 24h). If your remote base changes meaning within those 24 hours, it will not be detected unless a force refresh is performed.
> > We actually discourage this pattern where the remote base can change meaning since it is not gitops friendly (e.g. rollback is no longer deterministic).
> 
> When you say "we discourage this pattern" are you referring to the use of references to other github repos in kustomize files? If so, is there another method you suggest for avoiding a lot of copying between definitions for different environments?

The use of non-tagged references are discouraged. Ie, a kustomise base that isn’t use a commit sha/tag

**Comment by evgenii-rodionov-affise** [comment]
Can I change the default cache refresh interval? 

> This is current design. Argo CD caches the state of manifests by the commit hash (default 24h). If your remote base changes meaning within those 24 hours, it will not be detected unless a force refresh is performed.
> 
> We actually discourage this pattern where the remote base can change meaning since it is not gitops friendly (e.g. rollback is no longer deterministic).

Hi! Can I change the default cache refresh interval?
